File: src/skills.ts
import { createHash, randomUUID } from "node:crypto";
import { spawn } from "node:child_process";
import { cpSync, existsSync, lstatSync, mkdirSync, readdirSync, readFileSync, realpathSync, renameSync, rmSync } from "node:fs";
import { basename, relative, resolve, sep } from "node:path";
import { ToolBroker, type ToolOperation } from "./adapters.js";
import { redact } from "./audit.js";
import { builtins, type SkillResponse } from "./builtin-skills.js";
import { SkillError, ValidationError } from "./errors.js";
import type { JsonObject, JsonValue, SkillManifest, SkillRun } from "./models.js";
import { isRecord, parseSkillManifest, utcNow } from "./models.js";
import { Policy } from "./policy.js";
import { Workspace, type SkillLockRecord } from "./storage.js";

const manifestName = "skill.json";

function packageFiles(root: string, directory = root): string[] {
  const files: string[] = [];
  for (const entry of readdirSync(directory, { withFileTypes: true })) {
    const path = resolve(directory, entry.name);
    const stat = lstatSync(path);
    if (stat.isSymbolicLink()) throw new SkillError(`skill packages cannot contain symbolic links: ${relative(root, path)}`);
    if (stat.isDirectory()) {
      if (entry.name !== "__pycache__") files.push(...packageFiles(root, path));
    } else if (!entry.name.endsWith(".pyc") && !entry.name.endsWith(".pyo")) files.push(path);
  }
  return files.sort((a, b) => relative(root, a).localeCompare(relative(root, b)));
}

export function packageDigest(root: string): string {
  const hash = createHash("sha256");
  for (const file of packageFiles(root)) {
    const name = Buffer.from(relative(root, file)); const content = readFileSync(file);
    const nameLength = Buffer.alloc(4); nameLength.writeUInt32BE(name.length);
    const contentLength = Buffer.alloc(8); contentLength.writeBigUInt64BE(BigInt(content.length));
    hash.update(nameLength).update(name).update(contentLength).update(content);
  }
  return `sha256:${hash.digest("hex")}`;
}

export class SkillRegistry {
  constructor(readonly workspace: Workspace, readonly policy: Policy) {}

  loadManifest(packagePath: string): SkillManifest {
    try { return parseSkillManifest(JSON.parse(readFileSync(resolve(packagePath, manifestName), "utf8")) as unknown); }
    catch (error) { if (error instanceof ValidationError) throw error; throw new ValidationError(`invalid ${manifestName}: ${error instanceof Error ? error.message : String(error)}`); }
  }

  install(packagePath: string): { manifest: SkillManifest; digest: string; path: string } {
    const source = realpathSync(resolve(packagePath));
    if (!lstatSync(source).isDirectory()) throw new SkillError(`skill package is not a directory: ${source}`);
    packageFiles(source);
    const manifest = this.loadManifest(source); this.policy.checkInstall(manifest);
    const digest = packageDigest(source); const target = resolve(this.workspace.skills, manifest.id, manifest.version);
    this.workspace.initialize();
    if (existsSync(target)) {
      if (packageDigest(target) !== digest) throw new SkillError(`${manifest.id}@${manifest.version} is already installed with a different digest`);
    } else {
      mkdirSync(resolve(target, ".."), { recursive: true }); const temporary = `${target}.${randomUUID()}.tmp`;
      cpSync(source, temporary, { recursive: true, dereference: false }); renameSync(temporary, target);
    }
    const lock = this.workspace.getLock();
    lock.skills[manifest.id] = { version: manifest.version, digest, path: relative(this.workspace.root, target), source: manifest.source };
    this.workspace.saveLock(lock); return { manifest, digest, path: target };
  }

  list(): ({ id: string } & SkillLockRecord)[] { return Object.entries(this.workspace.getLock().skills).sort(([a], [b]) => a.localeCompare(b)).map(([id, value]) => ({ id, ...value })); }

  resolve(skillId: string): { manifest: SkillManifest; packagePath: string; digest: string } {
    const record = this.workspace.getLock().skills[skillId]; if (!record) throw new SkillError(`skill is not installed: ${skillId}`);
    const packagePath = resolve(this.workspace.root, record.path); const manifest = this.loadManifest(packagePath); const digest = packageDigest(packagePath);
    if (digest !== record.digest) throw new SkillError(`installed skill digest mismatch: ${skillId}`);
    return { manifest, packagePath, digest };
  }
}

function parseResponse(value: unknown): SkillResponse {
  if (!isRecord(value)) throw new SkillError("skill response must be an object");
  const output = value.output ?? {}; const evidence = value.evidence ?? []; const operations = value.operations ?? [];
  if (!isRecord(output)) throw new SkillError("skill output must be an object");
  if (!Array.isArray(evidence) || !evidence.every(isRecord)) throw new SkillError("skill evidence must be an array of objects");
  if (!Array.isArray(operations) || !operations.every((item) => isRecord(item) && typeof item.tool === "string" && isRecord(item.arguments))) throw new SkillError("skill operations must be valid tool operations");
  return { output: output as JsonObject, evidence: evidence as JsonObject[], operations: operations as ToolOperation[] };
}

export class SkillRuntime {
  constructor(readonly workspace: Workspace, readonly registry: SkillRegistry, readonly policy: Policy, readonly tools = new ToolBroker()) {}

  async run(skillId: string, payload: JsonObject, options: { changeId?: string; workspaceId?: string; approved?: boolean } = {}): Promise<SkillRun> {
    const { manifest, packagePath } = this.registry.resolve(skillId); this.policy.checkRun(manifest, options.approved ?? false);
    const id = `skill-run-${randomUUID().replaceAll("-", "")}`; const startedAt = utcNow();
    try {
      const response = parseResponse(await this.invoke(manifest, packagePath, payload));
      const toolResults: JsonObject[] = [];
      for (const operation of response.operations) toolResults.push(await this.tools.execute(operation, new Set(manifest.permissions)));
      const output = toolResults.length ? { ...response.output, toolResults } : response.output;
      const run: SkillRun = { id, skillId: manifest.id, skillVersion: manifest.version, ...(options.changeId ? { changeId: options.changeId } : {}), ...(options.workspaceId ? { workspaceId: options.workspaceId } : {}), status: "succeeded", startedAt, completedAt: utcNow(), input: redact(payload) as JsonObject, output: redact(output) as JsonObject, evidence: redact(response.evidence) as JsonObject[] };
      this.workspace.saveRun(run); return run;
    } catch (error) {
      const run: SkillRun = { id, skillId: manifest.id, skillVersion: manifest.version, ...(options.changeId ? { changeId: options.changeId } : {}), ...(options.workspaceId ? { workspaceId: options.workspaceId } : {}), status: "failed", startedAt, completedAt: utcNow(), input: redact(payload) as JsonObject, output: {}, evidence: [], error: `${error instanceof Error ? error.name : "Error"}: ${error instanceof Error ? error.message : String(error)}` };
      this.workspace.saveRun(run); throw error;
    }
  }

  private async invoke(manifest: SkillManifest, packagePath: string, payload: JsonObject): Promise<unknown> {
    if (manifest.entrypoint.type === "builtin") { const handler = builtins[manifest.entrypoint.target]; if (!handler) throw new SkillError(`unknown builtin skill: ${manifest.entrypoint.target}`); return handler(payload); }
    const executable = resolve(packagePath, manifest.entrypoint.target); const prefix = `${packagePath}${sep}`;
    if (!executable.startsWith(prefix) || !existsSync(executable) || lstatSync(executable).isDirectory()) throw new SkillError("process entrypoint must be a file inside its skill package");
    const extension = basename(executable).split(".").pop();
    const command = extension === "js" || extension === "mjs" ? process.execPath : extension === "py" ? "python3" : executable;
    const args = command === process.execPath || command === "python3" ? [executable] : [];
    return await new Promise((accept, reject) => {
      const child = spawn(command, args, { cwd: packagePath, env: { PATH: process.env.PATH ?? "", LANG: "C.UTF-8" }, stdio: ["pipe", "pipe", "pipe"] });
      let stdout = ""; let stderr = ""; let settled = false;
      const timeout = Math.min(manifest.timeoutSeconds, this.policy.processTimeoutSeconds);
      const timer = setTimeout(() => { child.kill("SIGKILL"); if (!settled) { settled = true; reject(new SkillError(`skill timed out after ${timeout} seconds`)); } }, timeout * 1000);
      child.stdout.setEncoding("utf8").on("data", (part: string) => { stdout += part; }); child.stderr.setEncoding("utf8").on("data", (part: string) => { stderr += part; });
      child.on("error", (error) => { clearTimeout(timer); if (!settled) { settled = true; reject(new SkillError(`skill process failed: ${error.message}`)); } });
      child.on("close", (code) => { clearTimeout(timer); if (settled) return; settled = true; if (code !== 0) return reject(new SkillError(`skill process exited ${code}: ${stderr.slice(-1000).trim()}`)); try { accept(JSON.parse(stdout) as unknown); } catch { reject(new SkillError("skill process did not return valid JSON")); } });
      child.stdin.end(JSON.stringify(payload));
    });
  }
}

File: src/cjhx_agile_workflow/skills.py
from __future__ import annotations

import hashlib
import json
import os
import shutil
import subprocess
import sys
import uuid
from pathlib import Path
from typing import Any

from .adapters import ToolBroker, ToolOperation
from .audit import redact
from .builtin_skills import BUILTINS
from .errors import SkillError, ValidationError
from .models import SkillManifest, SkillRun, utc_now
from .policy import Policy
from .storage import Workspace

MANIFEST_NAME = "skill.json"


def package_digest(path: Path) -> str:
    digest = hashlib.sha256()
    files = sorted(
        item
        for item in path.rglob("*")
        if item.is_file()
        and "__pycache__" not in item.parts
        and item.suffix not in {".pyc", ".pyo"}
    )
    for item in files:
        relative = item.relative_to(path).as_posix().encode()
        digest.update(len(relative).to_bytes(4, "big"))
        digest.update(relative)
        content = item.read_bytes()
        digest.update(len(content).to_bytes(8, "big"))
        digest.update(content)
    return f"sha256:{digest.hexdigest()}"


class SkillRegistry:
    def __init__(self, workspace: Workspace, policy: Policy) -> None:
        self.workspace = workspace
        self.policy = policy

    @staticmethod
    def load_manifest(package: Path) -> SkillManifest:
        manifest_path = package / MANIFEST_NAME
        try:
            value = json.loads(manifest_path.read_text(encoding="utf-8"))
        except FileNotFoundError as error:
            raise ValidationError(f"skill package requires {MANIFEST_NAME}: {package}") from error
        except json.JSONDecodeError as error:
            raise ValidationError(f"invalid {MANIFEST_NAME}: {error}") from error
        if not isinstance(value, dict):
            raise ValidationError("skill manifest must be a JSON object")
        return SkillManifest.from_dict(value)

    def install(self, package: str | Path) -> dict[str, Any]:
        source = Path(package).resolve()
        if not source.is_dir():
            raise SkillError(f"skill package is not a directory: {source}")
        manifest = self.load_manifest(source)
        symlinks = [item for item in source.rglob("*") if item.is_symlink()]
        if symlinks:
            raise SkillError(
                "skill packages cannot contain symbolic links: "
                + ", ".join(str(item.relative_to(source)) for item in symlinks)
            )
        self.policy.check_install(manifest)
        digest = package_digest(source)
        target = self.workspace.skills / manifest.id / manifest.version
        self.workspace.initialize()
        if target.exists():
            existing = package_digest(target)
            if existing != digest:
                raise SkillError(
                    f"{manifest.id}@{manifest.version} is already installed with a different digest"
                )
        else:
            target.parent.mkdir(parents=True, exist_ok=True)
            temporary = target.with_name(f".{target.name}.{uuid.uuid4().hex}.tmp")
            shutil.copytree(source, temporary)
            os.replace(temporary, target)
        lock = self.workspace.get_lock()
        lock["skills"][manifest.id] = {
            "version": manifest.version,
            "digest": digest,
            "path": str(target.relative_to(self.workspace.root)),
            "source": manifest.source,
        }
        self.workspace.save_lock(lock)
        return {"manifest": manifest.to_dict(), "digest": digest, "path": str(target)}

    def list(self) -> list[dict[str, Any]]:
        lock = self.workspace.get_lock()
        return [{"id": skill_id, **value} for skill_id, value in sorted(lock["skills"].items())]

    def resolve(self, skill_id: str) -> tuple[SkillManifest, Path, str]:
        lock = self.workspace.get_lock()
        record = lock["skills"].get(skill_id)
        if record is None:
            raise SkillError(f"skill is not installed: {skill_id}")
        package = self.workspace.root / record["path"]
        manifest = self.load_manifest(package)
        actual = package_digest(package)
        if actual != record["digest"]:
            raise SkillError(f"installed skill digest mismatch: {skill_id}")
        return manifest, package, actual


class SkillRuntime:
    def __init__(
        self,
        workspace: Workspace,
        registry: SkillRegistry,
        policy: Policy,
        tools: ToolBroker | None = None,
    ) -> None:
        self.workspace = workspace
        self.registry = registry
        self.policy = policy
        self.tools = tools or ToolBroker()

    def run(
        self,
        skill_id: str,
        payload: dict[str, Any],
        *,
        change_id: str | None = None,
        approved: bool = False,
    ) -> SkillRun:
        manifest, package, _ = self.registry.resolve(skill_id)
        self.policy.check_run(manifest, approved=approved)
        started = utc_now()
        run_id = f"skill-run-{uuid.uuid4().hex}"
        try:
            response = self._invoke(manifest, package, payload)
            output, evidence, operations = self._validate_response(response)
            tool_results = [
                self.tools.execute(ToolOperation.from_dict(item), set(manifest.permissions))
                for item in operations
            ]
            if tool_results:
                output = {**output, "toolResults": tool_results}
            run = SkillRun(
                id=run_id,
                skill_id=manifest.id,
                skill_version=manifest.version,
                change_id=change_id,
                status="succeeded",
                started_at=started,
                completed_at=utc_now(),
                input=redact(payload),
                output=redact(output),
                evidence=redact(evidence),
            )
        except Exception as error:
            run = SkillRun(
                id=run_id,
                skill_id=manifest.id,
                skill_version=manifest.version,
                change_id=change_id,
                status="failed",
                started_at=started,
                completed_at=utc_now(),
                input=redact(payload),
                output={},
                error=f"{type(error).__name__}: {error}",
            )
            self.workspace.save_run(run)
            raise
        self.workspace.save_run(run)
        return run

    def _invoke(
        self, manifest: SkillManifest, package: Path, payload: dict[str, Any]
    ) -> dict[str, Any]:
        if manifest.entrypoint.type == "builtin":
            handler = BUILTINS.get(manifest.entrypoint.target)
            if handler is None:
                raise SkillError(f"unknown builtin skill: {manifest.entrypoint.target}")
            return handler(payload)
        target = Path(manifest.entrypoint.target)
        if target.is_absolute() or ".." in target.parts:
            raise SkillError("process entrypoint must stay inside its skill package")
        executable = (package / target).resolve()
        if package.resolve() not in executable.parents or not executable.is_file():
            raise SkillError(f"process entrypoint not found: {target}")
        command = [sys.executable, str(executable)] if executable.suffix == ".py" else [str(executable)]
        timeout = min(manifest.timeout_seconds, self.policy.process_timeout_seconds)
        try:
            completed = subprocess.run(
                command,
                input=json.dumps(payload, ensure_ascii=False),
                capture_output=True,
                text=True,
                cwd=package,
                timeout=timeout,
                check=False,
                env={"PATH": os.environ.get("PATH", ""), "LANG": "C.UTF-8"},
            )
        except subprocess.TimeoutExpired as error:
            raise SkillError(f"skill timed out after {timeout} seconds") from error
        if completed.returncode != 0:
            detail = completed.stderr.strip()[-1000:]
            raise SkillError(f"skill process exited {completed.returncode}: {detail}")
        try:
            response = json.loads(completed.stdout)
        except json.JSONDecodeError as error:
            raise SkillError("skill process did not return valid JSON") from error
        if not isinstance(response, dict):
            raise SkillError("skill process response must be an object")
        return response

    @staticmethod
    def _validate_response(
        response: dict[str, Any],
    ) -> tuple[dict[str, Any], list[dict[str, Any]], list[dict[str, Any]]]:
        output = response.get("output", {})
        evidence = response.get("evidence", [])
        operations = response.get("operations", [])
        if not isinstance(output, dict):
            raise SkillError("skill output must be an object")
        if not isinstance(evidence, list) or not all(isinstance(item, dict) for item in evidence):
            raise SkillError("skill evidence must be an array of objects")
        if not isinstance(operations, list) or not all(isinstance(item, dict) for item in operations):
            raise SkillError("skill operations must be an array of objects")
        return output, evidence, operations
